handlers/handlers.go

```diff
@@ -28,86 +27,170 @@ type RenderClient interface {
 
 // SearchClient is an interface with methods required for a search client
 type SearchClient interface {
-	GetSearch(ctx context.Context, query url.Values) (r search.Response, err error)
+	GetSearch(ctx context.Context, query url.Values) (r searchC.Response, err error)
 }
 
-func setStatusCode(req *http.Request, w http.ResponseWriter, err error) {
-	status := http.StatusInternalServerError
-	if err, ok := err.(ClientError); ok {
-		if err.Code() == http.StatusNotFound {
-			status = err.Code()
-		}
+// Read Handler
+func Read(cfg *config.Config, rendC RenderClient, searchC SearchClient) http.HandlerFunc {
+	return func(w http.ResponseWriter, req *http.Request) {
+		read(w, req, cfg, rendC, searchC)
 	}
-	if err.Error() == "invalid filter type given" {
-		status = http.StatusBadRequest
+}
+
+func read(w http.ResponseWriter, req *http.Request, cfg *config.Config, rendC RenderClient, searchC SearchClient) {
+	ctx := req.Context()
+
+	urlQuery := req.URL.Query()
+
+	validatedQueryParams, err := data.ReviewQuery(ctx, cfg, urlQuery)
+	if err != nil {
+		log.Event(ctx, "unable to review query", log.Error(err), log.ERROR)
+		setStatusCode(w, req, err)
+		return
+	}
+
+	apiQuery := data.GetSearchAPIQuery(validatedQueryParams)
+
+	resp, err := searchC.GetSearch(ctx, apiQuery)
+	if err != nil {
+		logData := log.Data{"api query passed to search-api": apiQuery}
+		log.Event(ctx, "getting search response from client failed", log.Error(err), log.ERROR, logData)
+		setStatusCode(w, req, err)
+		return
+	}
+
+	// TO-DO: Until API handles aggregration on datatypes (e.g. bulletins, article), we need to make a second request
+
+	err = validateCurrentPage(ctx, validatedQueryParams, resp.Count)
+	if err != nil {
+		log.Event(ctx, "unable to validate current page", log.Error(err), log.ERROR)
+		setStatusCode(w, req, err)
+		return
+	}
+
+	categories, err := getCategoriesTypesCount(ctx, apiQuery, searchC)
+	if err != nil {
+		log.Event(ctx, "getting categories, types and its counts failed", log.Error(err), log.ERROR)
+		setStatusCode(w, req, err)
+		return
+	}
+
+	err = getSearchPage(w, req, cfg, rendC, validatedQueryParams, categories, resp)
+	if err != nil {
+		log.Event(ctx, "getting search page failed", log.Error(err), log.ERROR)
+		setStatusCode(w, req, err)
+		return
 	}
-	log.Event(req.Context(), "setting-response-status", log.Error(err), log.ERROR)
-	w.WriteHeader(status)
 }
 
-var marshal = func(v interface{}) ([]byte, error) {
-	return json.Marshal(v)
+// validateCurrentPage checks if the current page exceeds the total pages which is a bad request
+func validateCurrentPage(ctx context.Context, validatedQueryParams data.SearchURLParams, resultsCount int) (err error) {
```

Comment:
I wouldn't initialise err in function signature, so:

`func validateCurrentPage(ctx context.Context, validatedQueryParams data.SearchURLParams, resultsCount int) error {`

Then on line 93 add `:` to set err variable

That way it is following the same pattern as your other functions in this package.